FeedStoreChallenge/Core Data/CoreDataFeedStore.swift

```diff
@@ -0,0 +1,84 @@
+//
+//  Copyright © 2021 Essential Developer. All rights reserved.
+//
+
+import CoreData
+
+public final class CoreDataFeedStore: FeedStore {
+	private static let modelName = "FeedStore"
+	private static let model = NSManagedObjectModel(name: modelName, in: Bundle(for: CoreDataFeedStore.self))
+
+	private let container: NSPersistentContainer
+	private let context: NSManagedObjectContext
+
+	struct ModelNotFound: Error {
+		let modelName: String
+	}
+
+	public init(storeURL: URL) throws {
+		guard let model = CoreDataFeedStore.model else {
+			throw ModelNotFound(modelName: CoreDataFeedStore.modelName)
+		}
+
+		container = try NSPersistentContainer.load(
+			name: CoreDataFeedStore.modelName,
+			model: model,
+			url: storeURL
+		)
+		context = container.newBackgroundContext()
+	}
+    
+    private let queue = DispatchQueue(label: "\(CoreDataFeedStore.self)Queue", qos: .userInitiated, attributes: .concurrent)
+
+	public func retrieve(completion: @escaping RetrievalCompletion) {
+        queue.async { [unowned self] in
+            self.perform { context in
+                    do {
+                            if let cache = try ManagedCache.find(in: context) {
+                                completion(.found(feed: cache.localFeed, timestamp: cache.timestamp))
+                            } else {
+                               completion(.empty)
+                            }
+                        } catch {
+                            completion(.failure(error))
+                        }
+                }
+        }
```

Comment:
To follow up on the previous comment, please remove the `queue.async` call.

```suggestion
        perform { context in
                do {
                        if let cache = try ManagedCache.find(in: context) {
                            completion(.found(feed: cache.localFeed, timestamp: cache.timestamp))
                        } else {
                           completion(.empty)
                        }
                    } catch {
                        completion(.failure(error))
                    }
            }
    }
```